Allow use of custom HostnameVerifier on clients.

tbartley opened the pull request:
While the improvements to TLS configuration of HTTP clients in 1.0.0
(maybe prior) are awesome, as part of that process the ability to set a
custom HostnameVerifier easily on the HTTP client has been lost.

You used to be able to do e.g. as:

JerseyClientConfiguration myJerseyClientConfiguration = <some
configuration>;
HostnameVerifier verifier = new MyCustomHostnameVerifier();
JerseyClientBuilder clientBuilder = new JerseyClientBuilder(env);
clientBuilder.using(myJerseyClientConfiguration).using(verifier);
Client httpClient = clientBuilder.build();
Same is true for HttpClientBuilder too.

You can still do it by creating a custom Apache
Registry<ConnectionSocketFactory> but you need to set up socket
factories for every scheme.

This change restores the ability to set a custom HostnameVerifier
for clients.

[Fixes #1663]

tbartley commented:
Also, I wasn't sure how to approach writing a unit test in HttpClientBuilderTest.java. Some guidance there would be appreciated.

arteam commented on dropwizard-client/src/main/java/io/dropwizard/client/DropwizardSSLConnectionSocketFactory.java:
```diff
@@ -19,9 +19,16 @@
 
     public DropwizardSSLConnectionSocketFactory(TlsConfiguration configuration) {
         this.configuration = configuration;
+        this.verifier = null;
```
I think we could reduce this constructor to a call to another constructor:
```java
this(configuration, null);
```

tbartley replied:
Sure

arteam commented on dropwizard-client/src/main/java/io/dropwizard/client/DropwizardSSLConnectionSocketFactory.java:
```diff
@@ -48,7 +55,12 @@ public SSLConnectionSocketFactory getSocketFactory() throws SSLInitializationExc
         HostnameVerifier hostnameVerifier;
         if (configuration.isVerifyHostname()) {
-            hostnameVerifier = SSLConnectionSocketFactory.getDefaultHostnameVerifier();
+            if (verifier == null) {
```
What do you think about reducing the size of this method by using the ternary operator and several returns?
Something like that:
```java
if (configuration.isVerifyHostname()) {
    return verifier != null ? verifier : SSLConnectionSocketFactory.getDefaultHostnameVerifier();
} else {
    return new NoopHostnameVerifier();
}
```

tbartley replied:
Sure

arteam commented on dropwizard-client/src/main/java/io/dropwizard/client/HttpClientBuilder.java:
```diff
@@ -378,9 +396,15 @@ protected InstrumentedHttpClientConnectionManager createConnectionManager(Regist
         final SSLConnectionSocketFactory sslConnectionSocketFactory;
         if (configuration.getTlsConfiguration() == null) {
-            sslConnectionSocketFactory = SSLConnectionSocketFactory.getSocketFactory();
+            if (verifier == null) {
```
If I understand correctly, setting a `HostnameVerifier` means using a `TlsConfiguration`. What do you think about extracting `TlsConfiguration` to a local mutable variable?
```java
TlsConfiguration tlsConfiguration = configuration.getTlsConfiguration();
if (tlsConfiguration == null && verifier != null) {
    tlsConfiguration = new TlsConfiguration();
}

final SSLConnectionSocketFactory sslConnectionSocketFactory;
if (tlsConfiguration == null) {
    sslConnectionSocketFactory = SSLConnectionSocketFactory.getSocketFactory();
} else {
    sslConnectionSocketFactory = new DropwizardSSLConnectionSocketFactory(tlsConfiguration,
        verifier).getSocketFactory();
}
```

tbartley replied:
Sure

arteam commented on dropwizard-client/src/test/java/io/dropwizard/client/ssl/DropwizardSSLConnectionSocketFactoryTest.java:
```diff
@@ -172,4 +203,13 @@ public void shouldRejectNonSupportedProtocols() throws Exception {
+    private static class FailVerifier implements HostnameVerifier {
+
+		@Override
+		public boolean verify(String arg0, SSLSession arg1) {
```
Could you please replace tabs to spaces? A nitpick, but we prefer to stick to spaces across the project.

tbartley replied:
Ach! I'd been diligent about that elsewhere. Missed it here. Thanks.

arteam commented:
> Also, I wasn't sure how to approach writing a unit test in HttpClientBuilderTest.java. Some guidance there would be appreciated.

I think you could mark the method `createConfiguredRegistry` as package protected and put the `@VisibleForTesting` annotation on it. Then you could invoke it in tests:
```java
HostnameVerifier customVerifier = (s, sslSession) -> false;
Registry<ConnectionSocketFactory> configuredRegistry = builder
    .using(customVerifier)
    .createConfiguredRegistry();
```

Then we could retrieve the SSL socket factory from the registry and extract the verifier from it via reflection. Something like that:
```java
SSLConnectionSocketFactory socketFactory = 
(SSLConnectionSocketFactory) configuredRegistry.lookup("https");
final Field hostnameVerifierField =
    FieldUtils.getField(SSLConnectionSocketFactory.class, "hostnameVerifier", true);
```

tbartley commented:
Thanks for the feedback @arteam I've made your suggested changes and implemented additional unit test coverage following your suggested pattern.

I ummed and ahhed over the DropwizardSSLConnectionSocketFactory constructor unit test using reflection to check private field value but I liked using a getter for the verifier field on this class even less since it would imply getter for tlsConfiguration field for consistency which exposes that to internal modification which implies need for immutable TlsConfiguration .... Also, the getter would've needed to be public rather than package protected since unit test is in a different package.

I have also squashed my commits.

arteam commented:
Thank you very much for your contribution!

tbartley commented:
No problem at all. This software works so well for us. Thank you for the far greater contribution you make.
